### PR TITLE
Add algo for BooleanGateslogic

### PR DESCRIPTION
**Pull request by VANSH3104**
Adding algorithm logic for Logicgates like or and #5680 

- [ ] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [ ] This pull request is all my own work -- I have not plagiarized it.
- [ ] All filenames are in PascalCase.
- [ ] All functions and variable names follow Java naming conventions.
- [ ] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [ ] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

### CODE REVIEW

**Comment by saahil-mahato** [comment]
Please add tests and use proper javadoc standard found [here](https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html)

File: src/main/java/com/thealgorithms/bitmanipulation/BooleanAlgebraGates.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
+package com.thealgorithms.bitmanipulation;
+
+import java.util.List;
+
+/**
+ * Implements various Boolean algebra gates (AND, OR, NOT, XOR, NAND, NOR)
+ * using Java, generalized to handle multiple inputs.
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/bitmanipulation/BooleanAlgebraGates.java:7]: This "using Java" is redundant. Because it is already in Java repository

File: src/main/java/com/thealgorithms/bitmanipulation/BooleanAlgebraGates.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+ */
+public final class BooleanAlgebraGates {
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/bitmanipulation/BooleanAlgebraGates.java:9]: Looks good, but could you please add tests for this. They should be parametrized. A lot of examples are presented in project.

Possible tests:

```
1. AND Gate Tests:

Basic functionality:
AND([true, true]) should return true.
AND([true, false]) should return false.
AND([false, false]) should return false.

Multiple inputs:
AND([true, true, true]) should return true.
AND([true, false, true]) should return false.

Edge case (empty input list):
AND([]) can either return true (since the conjunction of no values is true in some interpretations) or throw an exception depending on your design.

2. OR Gate Tests:

Basic functionality:
OR([true, false]) should return true.
OR([false, false]) should return false.

Multiple inputs:
OR([true, true, false]) should return true.
OR([false, false, false]) should return false.

Edge case (empty input list):
OR([]) can either return false (since the disjunction of no values is false in some interpretations) or throw an exception.

3. NOT Gate Tests:
4. 
Basic functionality:
NOT(true) should return false.
NOT(false) should return true.

Edge case:
Not applicable, as NOT is a unary operation and requires a single input.

5. XOR Gate Tests:

Basic functionality:
XOR([true, false]) should return true.
XOR([true, true]) should return false.
XOR([false, false]) should return false.

Multiple inputs:
XOR([true, true, true]) should return true.
XOR([true, false, true]) should return false (since the XOR of an odd number of true inputs is true).

Edge case:
XOR([]) can either return false (if no inputs are given) or throw an exception.

7. NAND Gate Tests:

Basic functionality:
NAND([true, true]) should return false.
NAND([true, false]) should return true.
NAND([false, false]) should return true.

Multiple inputs:
NAND([true, true, true]) should return false.
NAND([true, true, false]) should return true.

Edge case:
NAND([]) can either return true or throw an exception depending on your design.

8. NOR Gate Tests:

Basic functionality:
NOR([true, false]) should return false.
NOR([false, false]) should return true.

Multiple inputs:
NOR([false, false, false]) should return true.
NOR([true, true, false]) should return false.

Edge case:
NOR([]) can either return true or throw an exception.

9. Edge Cases and Special Scenarios:

Empty input list:
Test how your program handles cases where an empty list is passed to gates expecting multiple inputs.
If the design assumes an empty list is invalid, it should throw an exception.

Single input for multi-input gates:
AND([true]) should return true.
OR([false]) should return false.
XOR([true]) should return true.

Test how gates behave with a single input as a corner case.

Mixed inputs:
For example, AND([true, false, false, true]) should return false, and similarly for other gates with mixed inputs.

10. Large Input Lists:

AND and OR gates:
Test how they perform with large input lists, e.g., AND([true, true, ..., true]) with 1,000 true values should return true.
Similarly, test OR with a large number of inputs containing mostly false and a single true.

12. Randomized Tests:

Generate random lists of true and false values and validate that the gates work as expected, especially for gates like XOR, which can have less obvious outputs for certain combinations.

13. Behavior on Invalid Input (if applicable):
Test what happens if null is passed as an input list or within the input list. The method should handle this either by throwing an exception or by some defined behavior.
```

**Review by alxkm — CHANGES_REQUESTED** [review]

**Comment by VANSH3104** [comment]
@alxkm not able to understand why i got clang format issue

**Comment by alxkm** [comment]
> @saahil-mahato @alxkm not able to understand why i got clang format issue

You can see details by clicking to the job details: https://github.com/TheAlgorithms/Java/actions/runs/11296030042/job/31420026439?pr=5717
Looks like there are redundant whitespaces.

But you added test descriptions to javadocs, which is wrong.
You should have added standard Java tests for the test cases I described in the review comment.

This is the implementation:

```
package com.thealgorithms.bitmanipulation;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Stream;
import com.thealgorithms.bitmanipulation.BooleanAlgebraGates.ANDGate;
import com.thealgorithms.bitmanipulation.BooleanAlgebraGates.BooleanGate;
import com.thealgorithms.bitmanipulation.BooleanAlgebraGates.NANDGate;
import com.thealgorithms.bitmanipulation.BooleanAlgebraGates.NORGate;
import com.thealgorithms.bitmanipulation.BooleanAlgebraGates.NOTGate;
import com.thealgorithms.bitmanipulation.BooleanAlgebraGates.ORGate;
import com.thealgorithms.bitmanipulation.BooleanAlgebraGates.XORGate;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.MethodSource;

class BooleanAlgebraGatesTest {

    @ParameterizedTest(name = "ANDGate Test Case {index}: inputs={0} -> expected={1}")
    @MethodSource("provideAndGateTestCases")
    void testANDGate(List<Boolean> inputs, boolean expected) {
        BooleanGate gate = new ANDGate();
        assertEquals(expected, gate.evaluate(inputs));
    }

    @ParameterizedTest(name = "ORGate Test Case {index}: inputs={0} -> expected={1}")
    @MethodSource("provideOrGateTestCases")
    void testORGate(List<Boolean> inputs, boolean expected) {
        BooleanGate gate = new ORGate();
        assertEquals(expected, gate.evaluate(inputs));
    }

    @ParameterizedTest(name = "NOTGate Test Case {index}: input={0} -> expected={1}")
    @CsvSource({
            "true, false",
            "false, true"
    })
    void testNOTGate(boolean input, boolean expected) {
        NOTGate gate = new NOTGate();
        assertEquals(expected, gate.evaluate(input));
    }

    @ParameterizedTest(name = "XORGate Test Case {index}: inputs={0} -> expected={1}")
    @MethodSource("provideXorGateTestCases")
    void testXORGate(List<Boolean> inputs, boolean expected) {
        BooleanGate gate = new XORGate();
        assertEquals(expected, gate.evaluate(inputs));
    }

    @ParameterizedTest(name = "NANDGate Test Case {index}: inputs={0} -> expected={1}")
    @MethodSource("provideNandGateTestCases")
    void testNANDGate(List<Boolean> inputs, boolean expected) {
        BooleanGate gate = new NANDGate();
        assertEquals(expected, gate.evaluate(inputs));
    }

    @ParameterizedTest(name = "NORGate Test Case {index}: inputs={0} -> expected={1}")
    @MethodSource("provideNorGateTestCases")
    void testNORGate(List<Boolean> inputs, boolean expected) {
        BooleanGate gate = new NORGate();
        assertEquals(expected, gate.evaluate(inputs));
    }

    // Helper methods to provide test data for each gate

    static Stream<Object[]> provideAndGateTestCases() {
        return Stream.of(
                new Object[]{Arrays.asList(true, true, true), true},
                new Object[]{Arrays.asList(true, false, true), false},
                new Object[]{Arrays.asList(false, false, false), false},
                new Object[]{Collections.emptyList(), true} // AND over no inputs is true
        );
    }

    static Stream<Object[]> provideOrGateTestCases() {
        return Stream.of(
                new Object[]{Arrays.asList(true, false, false), true},
                new Object[]{Arrays.asList(false, false, false), false},
                new Object[]{Arrays.asList(true, true, true), true},
                new Object[]{Collections.emptyList(), false} // OR over no inputs is false
        );
    }

    static Stream<Object[]> provideXorGateTestCases() {
        return Stream.of(
                new Object[]{Arrays.asList(true, false, true), false}, // XOR over odd true
                new Object[]{Arrays.asList(true, false, false), true},  // XOR over single true
                new Object[]{Arrays.asList(false, false, false), false},// XOR over all false
                new Object[]{Arrays.asList(true, true), false}          // XOR over even true
        );
    }

    static Stream<Object[]> provideNandGateTestCases() {
        return Stream.of(
                new Object[]{Arrays.asList(true, true, true), false},   // NAND of all true is false
                new Object[]{Arrays.asList(true, false), true},         // NAND with one false is true
                new Object[]{Arrays.asList(false, false), true},        // NAND of all false is true
                new Object[]{Collections.emptyList(), false}            // NAND over no inputs is false (negation of AND)
        );
    }

    static Stream<Object[]> provideNorGateTestCases() {
        return Stream.of(
                new Object[]{Arrays.asList(false, false), true},        // NOR of all false is true
                new Object[]{Arrays.asList(false, true), false},        // NOR with one true is false
                new Object[]{Arrays.asList(true, true), false},         // NOR of all true is false
                new Object[]{Collections.emptyList(), true}             // NOR over no inputs is true (negation of OR)
        );
    }
}
```

It may be necessary to correct some input data and the import order.

**Comment by VANSH3104** [comment]
@alxkm github action told that there is error I don't understand I check it in my local setup it don't show any error

**Review by alxkm — APPROVED** [review]
